### PR TITLE
 Allow 'bypass' keyword in a SQL, which runs the query directly at the backend database

### PR DESCRIPTION
**Pull request by dongyoungy**
Related to #202 and #194

### CODE REVIEW

File: src/main/java/org/verdictdb/jdbc41/VerdictStatement.java [hunk 1]
```diff
@@ -42,10 +42,67 @@ public VerdictStatement(VerdictConnection jdbcConn, VerdictContext context) {
...
+   */
+  private String checkBypass(String sql) {
+    if (sql.trim().toLowerCase().startsWith("bypass")) {
+      return sql.trim().substring(6);
+    }
+    return null;
+  }
+
+  private VerdictSingleResult executeAsIs(String sql) throws SQLException, VerdictDBDbmsException {
+    DbmsConnection dbmsConn = context.getConnection();
```
Comment (pyongjoo) [inline src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: Do we need this check? I guess a simple "dbmsConn.execute(sql)" could work?

File: src/main/java/org/verdictdb/jdbc41/VerdictStatement.java [hunk 2]
```diff
@@ -42,10 +42,67 @@ public VerdictStatement(VerdictConnection jdbcConn, VerdictContext context) {
     this.executionContext = context.createNewExecutionContext();
   }
 
+  /**
+   * Check whether given sql contains 'bypass' keyword at the beginning
+   *
+   * @param sql original sql
+   * @return without 'bypass' keyword if the original sql begins with it. null otherwise.
+   */
+  private String checkBypass(String sql) {
```
Comment (pyongjoo) [inline src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: Can we place this logic inside VerdictContext? The Java applications that directly use VerdictContext would not support BYPASS in this way.

Comment (pyongjoo) [reply src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: Sounds good to me.

File: src/main/java/org/verdictdb/jdbc41/VerdictStatement.java (hunk 1 above)
Comment (dongyoungy) [reply src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: Argh. I got it. Git highlighted checkByPass() method for the comment.

I guess you are right. I just wanted to be pedantic and use the actual JDBC statements whenever possible, but it does seem redundant at the same time.

**Comment by dongyoungy** [comment]
Changes as suggested have been pushed.

File: src/main/java/org/verdictdb/jdbc41/VerdictStatement.java [hunk 3]
```diff
@@ -45,7 +45,7 @@ public VerdictStatement(VerdictConnection jdbcConn, VerdictContext context) {
   @Override
   public boolean execute(String sql) throws SQLException {
     try {
-      result = executionContext.sql(sql);
+      result = context.sql(sql);
```
Comment (pyongjoo) [inline src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: I wonder if this is an intended behavior. My understanding was Connection:VerdictContext = Statement:ExecutionContext. If this was to handle bypass, I would suggest moving the logic into ExecutionContext, which still makes sense.

Comment (pyongjoo) [reply src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: It sounds good in general. But I still wonder how to close ExecutionContext when VerdictStatement.close() is called if VerdictStatement does not have the reference to ExecutionContext.

I think their conceptual relationship could be like, when a new VerdictStatement is created, VerdictConnection requests VerdictContext for a dedicated ExecutionContext, and then VerdictStatement simply uses the assigned ExecutionContext for query processing.

I agree that having a reference upper-level structure isn't good.

Comment (dongyoungy) [reply src/main/java/org/verdictdb/jdbc41/VerdictStatement.java]: Option 1: We can let VerdictContext to manage all of its ExecutionContext and have VerdictStatement to use its VerdictContext to close its associated ExecutionContext (e.g., using Map)

Option 2: We just have 1->2->4 and 3->4 execution path where every query processing logic goes into 4.

For now, I will use Option 2 as suggested while removing references to upper-level class/structures.

**Pull request merged**